### ISSUE TITLE
Differences of partial/row_wise/extract Optimizers

### ISSUE THREAD

**Issue by unavailableun** [pytorch/FBGEMM#1545]
Is there any document explaining the differences of below partial*, row_wise* and extract* optimizers?
Are they same with original optimizer from algorithm side? 

```Python
class EmbOptimType(enum.Enum):
    SGD = "sgd"  # uses non-deterministic updates (atomicAdd(..)) with duplicate ids
    EXACT_SGD = (
        "exact_sgd"  # uses deterministic updates (via sorting + segment reduction)
    )
    LAMB = "lamb"
    ADAM = "adam"
    # exact/dedup: gradients to the same row are applied with coalesce then apply
    # together, instead of applied in sequence (approx).
    EXACT_ADAGRAD = "exact_adagrad"
    EXACT_ROWWISE_ADAGRAD = "exact_row_wise_adagrad"
    LARS_SGD = "lars_sgd"
    PARTIAL_ROWWISE_ADAM = "partial_row_wise_adam"
    PARTIAL_ROWWISE_LAMB = "partial_row_wise_lamb"
    ROWWISE_ADAGRAD = "row_wise_adagrad"
    SHAMPOO = "shampoo"  # not currently supported for sparse embedding tables
    MADGRAD = "madgrad"
    EXACT_ROWWISE_WEIGHTED_ADAGRAD = "exact_row_wise_weighted_adagrad"
```

**Comment by jianyuh** [comment]
Please check Section 5 in https://arxiv.org/pdf/2104.05158.pdf . We plan to write more documents / paper for explanations.


exact_*  : 

As mentioned by comment, "exact" is also known as "dedup", where gradients to the same row are applied with coalesce then apply together, instead of applied in sequence (which is called "approx").

rowwise_* : 

The row-wise sparse AdaGrad was first introduced in [1], and first implemented in Caffe2 in https://caffe2.ai/docs/operators-catalogue.html#rowwisesparseadagrad. Let's assume we have a 𝐻 ×𝐷 embedding table. In the row-wise sparse AdaGrad, each element of the moment estimation is applied to the entire embedding row. For each row it is a single scaling factor that is updated by adding the average squared sum of gradients across the row. In this way, we keep the momentum as a 1D tensor with 𝐻 elements instead of 𝐻 ×𝐷 2D tensor, where 𝐻 and 𝐷 are the number of rows and the number of elements per row in an embedding table, respectively.

partial_rowwise_adam / partial_rowwise_lamb:

There are 2 optimizer states (momentums) in Adam/Lamb. We apply the row-wise to the 2nd optimizer state, and the keep the 1st optimizer state as a whole. This is for keep accuracy.


[1] Maya R Gupta, Samy Bengio, and Jason Weston. 2014. Training highly multiclass classifiers. The Journal of Machine Learning Research 15, 1 (2014), 1461–1492.


CC @sryap

**Comment by unavailableun** [comment]
Thanks @jianyuh for your explanation!
